opexebo/analysis/gridscore.py

```diff
@@ -148,124 +147,191 @@ def gridscore(aCorr, **kwargs):
-        
         # Where two fields have a very similar orientation, discard the more distant one
         orient_distsq = np.abs(_circ_dist2(orientation))
-        close_fields = orient_distsq < min_orientation  
+        close_fields = orient_distsq < min_orientation
         close_fields = np.triu(close_fields, 1) # Upper triangle only - set lower triangle to zero
                                                 # k=1: +1 offset from diagonal: set diagonal to zero too
-        row, col = np.where(close_fields)
-        to_del = np.zeros(row.size) # this is more efficient than appending to a list
-        for i in range(row.size):
-            r = row[i]
-            c = col[i]
-            if distance[r] > distance[c]:
-                j = r
+        to_del = []
+        for row,col in np.argwhere(close_fields):
+            if distance[row] > distance[col]:
+                to_del.append(row)
             else:
-                j = c
-            to_del[i] = int(j)
-        to_del = np.unique(to_del).astype(int) # this gets rid of all the zeros
-        distance = np.delete(distance, to_del)
-        all_coords = np.delete(all_coords, to_del, axis=0)
+                to_del.append(col)
+
+        distance    = np.delete(distance, to_del)
         orientation = np.delete(orientation, to_del)
-        # Actual deletion instead of substituting values - this handles cases 
-        # where there are fewer than 6 fields remaining after filtering        
-        
-        # Consider the 6 closest fields -> sort by distance
-        # The central point has already been deleted.
-        # Positions needs all 6 indicies, in order to calculate the ellipse
-        # Discard the duplicate values -> because they're floating point, don't use np.unique()
-        # Instead, use array slicing: sort by size, and then take every 2nd value
-        # Numpy slicing: [start:stop:step]
-        sorted_ids = np.argsort(distance)[:6]
-        positions = all_coords[sorted_ids,:]
-        gs_spacing = distance[sorted_ids[::2]]
-        gs_orientation = orientation[sorted_ids[::2]]
-        gs_orientation = np.degrees(gs_orientation)%180 # convert from rad to deg and wrap into [0-180]°
-        
+        all_coords  = np.delete(all_coords, to_del, axis=0)
+
+        # First sort by distance and take first 6 fields
+        sorted_ids_dist = np.argsort(distance)[:6] # 6 closest fields
+        positions, distance, orientation = all_coords[sorted_ids_dist], distance[sorted_ids_dist], orientation[sorted_ids_dist]
+        # ... then re-sort remaining fields by angle
+        sorted_ids_ang = np.argsort(orientation)
+
+        ################# GATHER OUTPUT #################
+        gs_positions = positions[sorted_ids_ang]
+        # For grid orientation and spacing take only 3 out of 6 neighbouring fields
+        gs_spacings = distance[sorted_ids_ang][:3]
+        gs_orientations = np.degrees(orientation[sorted_ids_ang][:3]) % 180
+
```

Comment:
How do you exclude the central field at dist=0?

Reply:
I am masking it before I detect fields (mask is created in line 153 of gridscore.py for both > best grid score radius and < center field radius).